middleware/jwt.go

```diff
@@ -15,22 +15,76 @@ import (
 	"errors"
 	"fmt"
 	"net/http"
+	"os"
+	"strconv"
 	"strings"
+	"time"
 
 	"github.com/auth0/go-jwt-middleware"
 	jwt "github.com/dgrijalva/jwt-go"
 	"github.com/gorilla/context"
+	"github.com/heketi/heketi/pkg/utils"
 )
 
+var (
+	logger          = utils.NewLogger("[jwt]", utils.LEVEL_DEBUG)
+	iatLeeway int64 = 120
+)
+
+func init() {
+	if leeway, exists := os.LookupEnv("HEKETI_JWT_IAT_LEEWAY_SECONDS"); exists {
+		value, err := strconv.Atoi(leeway)
+		if err != nil {
+			logger.Info("HEKETI_JWT_IAT_LEEWAY_SECONDS not valid: err, input string: %v, using default: %v", err, leeway, 120)
+		} else {
+			iatLeeway = int64(value)
+		}
+	}
+}
+
 // From https://github.com/dgrijalva/jwt-go/pull/139 it is understood
 // that if the machine where jwt token is generated and/or the machine
 // where jwt token is verified have any clock skew then there is a
-// possibility of getting a "Token used before issued" error. Considering
-// that we also check for expiration with delta of 5 minutes, disabling
-// iat claim until the patch is merged in jwt.
-var (
-	required_claims = []string{"iss", "exp"}
-)
+// possibility of getting a "Token used before issued" error.
+// Therefore we have implemented a derived claim type of HeketiJwtClaims and
+// the validation function for "iat" compares time with provision for leeway.
+// Also, now that we control the validation function, we make "exp" validation
+// mandatory and get rid of required_claims array.
+type HeketiJwtClaims struct {
+	*jwt.StandardClaims
+	Qsh string `json:"qsh,omitempty"`
+}
+
+func (c *HeketiJwtClaims) Valid() error {
+	vErr := new(jwt.ValidationError)
+	now := jwt.TimeFunc().Unix()
+
+	if c.VerifyExpiresAt(now, true) == false {
+		delta := time.Unix(now, 0).Sub(time.Unix(c.ExpiresAt, 0))
+		vErr.Inner = fmt.Errorf("Token is expired by %v", delta)
+		vErr.Errors |= jwt.ValidationErrorExpired
+		logger.LogError("exp validation failed: %v", vErr.Error())
+	}
+
+	// "iat" check
+	if now < c.IssuedAt-iatLeeway {
+		vErr.Inner = fmt.Errorf("Token used before issued")
+		vErr.Errors |= jwt.ValidationErrorIssuedAt
+		logger.LogError("iat validation failed: %v, time now: %v, time issued: %v", vErr.Error(), now, c.IssuedAt)
```

Comment:
```IssuedAt``` and ```now``` are both in Unix form.
can we convert back the Unix time to UTC time for more readability on logging

Reply:
done